File: 8CircularQueue.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#define size 5 
// Implemenatation of circular queue using circular array.
// Circular array is the linear array with modular arithmetic so that no space is wasted.
// Overflow condition=> when front==(rear+1) mod size
// Underflow condition=> when front==rear

int queue[size];
int front=0,rear=0; 

// function  to enqueue in the queue.
void enqueue(int data){
rear=(rear+1)%size;
// first checking the overflow condition
if(front==rear){
    printf("Queue is full ");
    // have to make rear node pointing to the last element(as we already incremented)
    if(rear==0) 
       rear=size-1;
    else
      rear--;  
    
    }
    // if no overflow store the data
else{
    queue[rear]=data;
    }
}

void dequeue(){
    //checking the underflow condition
    if(front==rear){
        printf("Queue is empty");
    }
    else{
        //dequeuing the data from the front node.
        front=(front+1)%size;
        printf("%d ",queue[front]);
    }
}
// Main program
int main(){
    enqueue(1);
    enqueue(2);
    enqueue(3);
    enqueue(4);
    dequeue();
    dequeue();
    dequeue();
    dequeue();
}

File: 8CircularQueue.py
##  Implemenatation of circular queue using circular array.
##  Circular array is the linear array with modular arithmetic so that no space is wasted.
##  Overflow condition=> when front==(rear+1) mod size
##  Underflow condition=> when front==rear

## This class will contains all the basic operation of CircularQueue
class CircularQueue:
    ## size parameter should be given at the initiation
    def __init__(self,size): 
        self.front=0
        self.rear=0
        self.queue=[0]*size 
        self.size=size
        
    def enqueue(self,data):
        self.rear=(self.rear+1)%self.size
        if(self.front==self.rear): ## checking overflow condition
            print("Overflow")
            if(self.rear==0):
                self.rear=self.size-1
            else:
                self.rear-=1
            return
        else:
            self.queue[self.rear]=data 
        
    def dequeue(self):
        if(self.front==self.rear): ## checking underflow condition
            print("Queue is empty")
        else:
            self.front+=1
            print(self.queue[self.front])
  
## main program
if __name__=="__main__":
    A=CircularQueue(5)
    A.enqueue(1)
    A.enqueue(2)
    A.enqueue(3)
    A.enqueue(4)
    A.dequeue()
    A.dequeue()
    A.dequeue()
    A.dequeue()
    
        
